json.go

```diff
@@ -0,0 +1,367 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/checkpoint-restore/go-criu/v7/crit"
+	spec "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+type CheckpointSize struct {
+	TotalSize             int64 `json:"total_size,omitempty"`
+	MemoryPagesSize       int64 `json:"memory_pages_size,omitempty"`
+	AmdGpuMemoryPagesSize int64 `json:"amd_gpu_memory_pages_size,omitempty"`
+	RootFsDiffSize        int64 `json:"root_fs_diff_size,omitempty"`
+}
+
+type StatsNode struct {
+	FreezingTime uint32 `json:"freezing_time,omitempty"`
+	FrozenTime   uint32 `json:"frozen_time,omitempty"`
+	MemdumpTime  uint32 `json:"memdump_time,omitempty"`
+	MemwriteTime uint32 `json:"memwrite_time,omitempty"`
+	PagesScanned uint64 `json:"pages_scanned,omitempty"`
+	PagesWritten uint64 `json:"pages_written,omitempty"`
+}
+
+type PsNode struct {
+	PID      uint32            `json:"pid"`
+	Comm     string            `json:"command"`
+	Cmdline  string            `json:"cmdline,omitempty"`
+	EnvVars  map[string]string `json:"environment_variables,omitempty"`
+	Children []PsNode          `json:"children,omitempty"`
+}
+
+type FdNode struct {
+	PID       uint32         `json:"pid"`
+	OpenFiles []OpenFileNode `json:"open_files,omitempty"`
+}
+
+type OpenFileNode struct {
+	Type string `json:"type"`
+	FD   string `json:"fd"`
+	Path string `json:"path"`
+}
+
+type SkNode struct {
+	PID         uint32       `json:"pid"`
+	OpenSockets []SocketNode `json:"open_sockets,omitempty"`
+}
+
+type SocketNode struct {
+	Protocol string `json:"protocol"`
+	Data     string `json:"data"`
```

Comment:
It might be good to also define a type socket data.

Example:
```Go
type SkData struct {
	Type       string `json:"type,omitempty"`
	Address    string `json:"address,omitempty"`
	State      string `json:"state,omitempty"`
	Source     string `json:"src,omitempty"`
	SourcePort uint32 `json:"src_port,omitempty"`
	Dest       string `json:"dst,omitempty"`
	DestPort   uint32 `json:"dst_port,omitempty"`
	SendBuf    string `json:"send_buf,omitempty"`
	RecvBuf    string `json:"recv_buf,omitempty"`
}

type SocketNode struct {
	Protocol string `json:"protocol,omitempty"`
	Data     SkData `json:"data,omitempty"`
}

var socketDataFuncs = map[string]func(*crit.Socket) (SkData, error){
	"UNIXSK":    getUnixSkData,
	"INETSK":    getInetSkData,
	"PACKETSK":  getPacketSkData,
	"NETLINKSK": getNetlinkSkData,
}

...

func buildJSONSks(sks []*crit.Sk) ([]SkNode, error) {
	var result []SkNode

	for _, sk := range sks {
		skNode := SkNode{
			PID: sk.PId,
		}

		var sockets []SocketNode
		for _, socket := range sk.Sockets {
			socketData, err := getDataForSocket(socket)
			if err != nil {
				return nil, fmt.Errorf("error getting data for socket: %w", err)
			}
			sockets = append(sockets, SocketNode{
				Protocol: socket.Protocol,
				Data:     socketData,
			})
		}

		if len(sockets) > 0 {
			skNode.OpenSockets = sockets
		}

		result = append(result, skNode)
	}

	return result, nil
}

func getDataForSocket(socket *crit.Socket) (SkData, error) {
	dataFunc, found := socketDataFuncs[socket.FdType]
	if !found {
		return SkData{}, fmt.Errorf("unsupported socket type: %s", socket.FdType)
	}
	return dataFunc(socket)
}

func getUnixSkData(socket *crit.Socket) (SkData, error) {
	return SkData{
		Type:    "UNIX",
		Address: socket.SrcAddr,
	}, nil
}

func getInetSkData(socket *crit.Socket) (SkData, error) {
	return SkData{
		Type:       socket.Protocol,
		State:      socket.State,
		Source:     socket.SrcAddr,
		SourcePort: socket.SrcPort,
		Dest:       socket.DestAddr,
		DestPort:   socket.DestPort,
		SendBuf:    socket.SendBuf,
		RecvBuf:    socket.RecvBuf,
	}, nil
}

func getPacketSkData(socket *crit.Socket) (SkData, error) {
	return SkData{
		SendBuf: socket.SendBuf,
		RecvBuf: socket.RecvBuf,
	}, nil
}

func getNetlinkSkData(socket *crit.Socket) (SkData, error) {
	return SkData{
		SendBuf: socket.SendBuf,
		RecvBuf: socket.RecvBuf,
	}, nil
}
```

Reply:
yeah sure,